sourcers/json/json.go

```diff
@@ -0,0 +1,62 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/ardanlabs/conf"
+)
+
+type JSONSourcer struct {
+	m map[string]string
+}
+
+// NewSource returns a conf.Sourcer and, potentially, an error if a
+// read error occurs or the Reader contains an invalid JSON document.
+func NewSource(r io.Reader) (conf.Sourcer, error) {
+	if r == nil {
+		return &JSONSourcer{m: nil}, nil
+	}
+
+	src, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	tmpMap := make(map[string]interface{})
+	err = json.Unmarshal(src, &tmpMap)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]string)
+	for key, value := range tmpMap {
+		switch v := value.(type) {
+		case float64:
+			m[key] = strings.TrimRight(fmt.Sprintf("%f", v), "0.")
+		case bool:
+			m[key] = fmt.Sprintf("%t", v)
+		case string:
+			m[key] = value.(string)
+		}
+	}
+
+	return &JSONSourcer{m: m}, nil
+}
+
+func (s *JSONSourcer) Source(fld conf.Field) (string, bool) {
```

Comment:
Let's use a better receiver name. The letter s it's too overused for strings and other things.

Maybe the problem is the redudancy in the name JSONSourcer. That's not good. Plus the "er" suffix should be reserved for interfaces.

If the type was named Source, it would remove the redudnancy `json.Source`,  `json.NewSource`,  that's better and can be used across all the packages.

then you would have
`func (src *Source) Source(fld conf.Field) (string, bool) {`

There are problems with the above as well

`json.JSON`, `json.New()`
`func (j *JSON) Source(fld conf.Field) (string, bool) {`

I don't think the user will be importing the `encoding\json` package at the same time they are importing the sourcer, so there won't be a name clash. Your test kind of proves that.

We need to get this right with this package before looking at the others.